server/server.js

```diff
@@ -142,24 +146,49 @@ Server.prototype._handleClientMessage = function(client, data) {
     return;
   }
 
-  logging.info('#client.message - received', client.id, message,
-     (this.resources[message.to] ? 'exists' : 'not instantiated'),
-     (this.subs[message.to] ? 'is subscribed' : 'not subscribed')
-    );
+  if (!this._clientDataPersist(client.id, message)) {
+    return;
+  }
+
+  this._resourceMessageHandle(client, message);
+};
+
+// Initialize a client, and persist messages where required
+Server.prototype._clientDataPersist = function (id, message) {
+  // Sync the client name to the current client id
+  if (message.op == 'name_id_sync') {
+    this._clientInit(message);
+    this.clientVersion = message.client_version;
+  }
+  else if (this.clientVersion &&
+        Semver.gte(this.clientVersion, VERSION_CLIENT_DATASTORE) &&
+        !Client.dataStore(id, message)) {
+    return false;
+  }
+
+  return true;
+};
 
+// Get a resource, subscribe where required, and handle associated message
+Server.prototype._resourceMessageHandle = function (client, message) {
   var resource = this._resourceGet(message.to);
   if (resource) {
+    logging.info('#client.message - received', client.id, message,
+      (this.resources[message.to] ? 'exists' : 'not instantiated'),
+      (this.subs[message.to] ? 'is subscribed' : 'not subscribed')
+      );
+
     this._persistenceSubscribe(resource.name, client.id)
     resource.handleMessage(client, message);
     this.emit(message.op, client, message);
   }
 };
 
-// Authorize client message
-Server.prototype._messageAuthorize =  function (message, client, data) {
+// Authorize a client message
+Server.prototype._messageAuthorize =  function (message, client) {
   var rtn = Core.Auth.authorize(message, client);
   if (!rtn) {
```

Comment:
rtn is a general name I use for "return value", but explicit naming is certainly better in this case.